tests/zfs-tests/tests/functional/zap_shrink/zap_shrink_001_pos.ksh

```diff
@@ -0,0 +1,83 @@
+#! /bin/ksh -p
+#
+# CDDL HEADER START
+#
+# The contents of this file are subject to the terms of the
+# Common Development and Distribution License (the "License").
+# You may not use this file except in compliance with the License.
+#
+# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+# or https://opensource.org/licenses/CDDL-1.0.
+# See the License for the specific language governing permissions
+# and limitations under the License.
+#
+# When distributing Covered Code, include this CDDL HEADER in each
+# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+# If applicable, add the following below this CDDL HEADER, with the
+# fields enclosed by brackets "[]" replaced with your own identifying
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright 2024, Klara Inc.
+#
+
+. $STF_SUITE/include/libtest.shlib
+
+#
+# DESCRIPTION:
+# Create a large number of files in a directory. Then remove all files and
+# check that the directory zap was shrunk. Use zdb to check that the zap object
+# contains only one leaf block using zdb.
+#
+
+verify_runnable "global"
+
+function cleanup
+{
+	rm -rf $TESTDIR/xxx
+}
+
+function check_zap_shrunk
+{
+	dir=$1
+	res=$(zdb -v -O "$TESTPOOL/$TESTFS" $dir | grep "Leaf blocks:" | awk -F\: '{print($2);}')
+	test 1 -eq $res
+}
+
+NR_FILES=100000
+BATCH=1000
+CWD=$PWD
+
+log_assert "Create a large number of files ($NR_FILES) in a directory. " \
+	"Make sure that the directory ZAP object was shrunk."
+
+log_onexit cleanup
+
+log_must mkdir $TESTDIR/xxx
+
+cd $TESTDIR/xxx
+# In order to prevent arguments overflowing, create NR_FILES in BATCH at once.
+for i in $(seq $(($NR_FILES/$BATCH))); do
+	touch $(seq $((($i-1)*$BATCH+1)) $(($i*$BATCH)));
+done
+cd $CWD
+
+log_must test $NR_FILES -eq $(ls -U $TESTDIR/xxx | wc -l)
+
+# remove all files in xxx directory
+cd $TESTDIR/xxx
+for i in $(seq $(($NR_FILES/$BATCH))); do
+	rm -f $(seq $((($i-1)*$BATCH+1)) $(($i*$BATCH)))
+done
+cd $CWD
+
+log_must test 0 -eq $(ls -U $TESTDIR/xxx | wc -l)
+
+# check whether zap_shrink works
+sleep 7 # wait for the txg synced (ugly, but what is a better way?)
+log_must check_zap_shrunk xxx
```

Comment:
We should probably export the pool, check the zap, then reimport the pool.  This way the pool can't change and we won't have to worry about `zdb` failing unexpectedly.

Reply:
done